pkg/yurttunnel/dns/dns.go

```diff
@@ -0,0 +1,463 @@
+		options.FieldSelector = selector
+	}
+	return coreinformers.NewFilteredServiceInformer(cs, constants.YurttunnelServerServiceNs, resyncPeriod, nil, tweakListOptions)
+}
+
+// newConfigMapInformer creates a shared index informer that returns only interested configmaps
+func newConfigMapInformer(cs clientset.Interface, resyncPeriod time.Duration) cache.SharedIndexInformer {
+	selector := fmt.Sprintf("metadata.name=%v", util.YurttunnelServerDnatConfigMapName)
+	tweakListOptions := func(options *metav1.ListOptions) {
+		options.FieldSelector = selector
+	}
+	return coreinformers.NewFilteredConfigMapInformer(cs, util.YurttunnelServerDnatConfigMapNs, resyncPeriod, nil, tweakListOptions)
+}
+
+func (dnsctl *coreDNSRecordController) Run(stopCh <-chan struct{}) {
+	electionChecker := leaderelection.NewLeaderHealthzAdaptor(time.Second * 20)
+	id, err := os.Hostname()
+	if err != nil {
+		klog.Fatalf("failed to get hostname, %v", err)
+	}
+	rl, err := resourcelock.New("leases", metav1.NamespaceSystem, "tunnel-dns-controller",
+		dnsctl.kubeClient.CoreV1(),
+		dnsctl.kubeClient.CoordinationV1(),
+		resourcelock.ResourceLockConfig{
+			Identity: id + "_" + string(uuid.NewUUID()),
+		})
+	if err != nil {
+		klog.Fatalf("error creating tunnel-dns-controller lock, %v", err)
+	}
+
+	leaderelection.RunOrDie(context.TODO(), leaderelection.LeaderElectionConfig{
+		Lock:          rl,
+		LeaseDuration: metav1.Duration{Duration: time.Second * time.Duration(15)}.Duration,
+		RenewDeadline: metav1.Duration{Duration: time.Second * time.Duration(10)}.Duration,
+		RetryPeriod:   metav1.Duration{Duration: time.Second * time.Duration(2)}.Duration,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				dnsctl.run(stopCh)
+			},
+			OnStoppedLeading: func() {
+				klog.Fatalf("leaderelection lost")
+			},
+		},
+		WatchDog: electionChecker,
+		Name:     "tunnel-dns-controller",
+	})
+	panic("unreachable")
+}
+
+func (dnsctl *coreDNSRecordController) run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer dnsctl.queue.ShutDown()
+
+	klog.Infof("starting tunnel dns controller")
+	defer klog.Infof("shutting down tunnel dns controller")
+
+	if !cache.WaitForNamedCacheSync("tunnel-dns-controller", stopCh,
+		dnsctl.nodeListerSynced, dnsctl.svcInformerSynced, dnsctl.cmInformerSynced) {
+		return
+	}
+
+	if err := dnsctl.ensureCoreDNSRecordConfigMap(); err != nil {
+		klog.Errorf("failed to ensure dns record ConfigMap %v/%v, %v",
+			yurttunnelDNSRecordConfigMapNs, yurttunnelDNSRecordConfigMapName, err)
+		return
+	}
+
+	go wait.Until(dnsctl.worker, time.Second, stopCh)
+
+	// sync dns hosts as a whole
+	go wait.Until(dnsctl.syncDNSRecordAsWhole, time.Duration(dnsctl.syncPeriod)*time.Second, stopCh)
+
+	// sync tunnel server svc
+	go wait.Until(func() {
+		if err := dnsctl.syncTunnelServerService(); err != nil {
+			klog.Errorf("failed to sync tunnel server service, %v", err)
+		}
```

Comment:
why we need to add syncTunnelServerService here?

Reply:
Based on past experience, we should not rely solely on the `watch` mechanism, as there is a risk of lost events. We need to make sure that the information is synchronized through a periodic `list` function.